### PR TITLE
add support for instantiating lonely manifest

### PR DESCRIPTION
**Pull request by KristofferC**
Fixes https://github.com/JuliaLang/Pkg.jl/issues/1091.

~~Not sure if the presence of a manifest file in the current directory is the best way to activate this feature or if we should enforce a flag to instantiate to do it.~~

### CODE REVIEW

**Comment by fredrikekre** [comment]
> Not sure if the presence of a manifest file in the current directory is the best way to activate this feature or if we should enforce a flag to instantiate to do it.

My thinking was that we do this if the project is activated and there is no project file.

**Comment by KristofferC** [comment]
Ah, makes sense.

**Comment by KristofferC** [comment]
How about this?

File: src/API.jl [hunk 1]
```diff
@@ -709,6 +709,22 @@ instantiate(; kwargs...) = instantiate(Context(); kwargs...)
...
+        _manifest = Pkg.Types.read_manifest(manifestfile)
+        deps = Dict()
+        for (uuid, pkg) in _manifest
+            if pkg.name in keys(deps)
+                # TODO, query what package to put in Project when in interactive mode?
+                pkgerror("cannot instantiate a manifest without project file when the manifest has multiple packages with the same name ($(pkg.name))")
+            end
+            deps[pkg.name] = string(uuid)
+        end
+        Types.write_project(Dict("deps" => deps), "Project.toml")
```
Comment (fredrikekre) [inline src/API.jl]: The project might not be in the currect directory:
```suggestion
        Types.write_project(Dict("deps" => deps), Types.find_project_file())
```

Comment (fredrikekre) [reply src/API.jl]: Or I guess:
```suggestion
        Types.write_project(Dict("deps" => deps), ctx.env.project_file)
```

Comment (KristofferC) [reply src/API.jl]: Yeah, I accidentally left some of that assumption in, thanks.

File: src/API.jl [hunk 2]
```diff
@@ -709,6 +709,22 @@ instantiate(; kwargs...) = instantiate(Context(); kwargs...)
 function instantiate(ctx::Context; manifest::Union{Bool, Nothing}=nothing,
                      update_registry::Bool=true, verbose::Bool=false, kwargs...)
     Context!(ctx; kwargs...)
+    if !isfile(ctx.env.project_file) && isfile(ctx.env.manifest_file)
+        if manifest == false
+            pkgerror("cannot instantiate without a project file with `manifest=false` option")
+        end
+        _manifest = Pkg.Types.read_manifest(manifestfile)
```
Comment (fredrikekre) [inline src/API.jl]:
```suggestion
        _manifest = Pkg.Types.read_manifest(ctx.env.manifest_file)
```

**Comment by KristofferC** [comment]
bor try

**Comment by KristofferC** [comment]
bors try

**Comment by KristofferC** [comment]
bors retry

File: src/Pkg.jl [hunk 3]
```diff
@@ -311,12 +311,14 @@ const project = API.project
...
 
-If a `Manifest.toml` file exists in the current project, download all
+If a `Manifest.toml` file exists in the active project, download all
 the packages declared in that manifest.
 Otherwise, resolve a set of feasible packages from the `Project.toml` files
 and install them.
 `verbose = true` prints the build output to `stdout`/`stderr` instead of
 redirecting to the `build.log` file.
+If no `Project.toml` exist in the current active project, create one with all the
+dependencies in the manifest.
```
Comment (00vareladavid) [inline src/Pkg.jl]:
```suggestion
dependencies in the manifest, then install all the packages declared in the manifest.
```
Perhaps this is more clear that you proceed with a normal `instantiate` after the project file is created.

File: src/API.jl [hunk 4]
```diff
@@ -709,6 +709,22 @@ instantiate(; kwargs...) = instantiate(Context(); kwargs...)
...
+        deps = Dict()
+        for (uuid, pkg) in _manifest
+            if pkg.name in keys(deps)
+                # TODO, query what package to put in Project when in interactive mode?
+                pkgerror("cannot instantiate a manifest without project file when the manifest has multiple packages with the same name ($(pkg.name))")
+            end
+            deps[pkg.name] = string(uuid)
+        end
+        Types.write_project(Dict("deps" => deps), ctx.env.project_file)
+        return instantiate(Context(); manifest=manifest, update_registry=update_registry, verbose=verbose)
```
Comment (fredrikekre) [inline src/API.jl]:
```suggestion
        return instantiate(Context(); manifest=manifest, update_registry=update_registry, verbose=verbose, kwargs...)
```

Comment (KristofferC) [reply src/API.jl]: The kwargs are already applied to the context though?

Comment (00vareladavid) [reply src/API.jl]: But this line creates a new context as well, it doesn't seem that e.g. `io=stdout` would go through

Comment (KristofferC) [reply src/API.jl]: Oh yeah...

**Pull request merged**